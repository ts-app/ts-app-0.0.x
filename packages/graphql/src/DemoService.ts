import { Resolver } from './Resolver'
import { ServiceInfo } from './ServiceInfo'

export class DemoService {
  private anotherNumber: number = 123

  constructor (public count: number = 0) {
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}"`
  }

  @Resolver({ type: 'mutation' })
  add ({ howMany = 1 }: { howMany?: number }) {
    this.count += howMany
    return this.count
  }

  @Resolver({ type: 'mutation' })
  async promiseToSubtract ({ howMany = 1 }: { howMany?: number }): Promise<number> {
    return new Promise<number>(resolve => {
      setTimeout(() => {
        this.count -= howMany
        resolve(this.count)
      }, 1)
    })
  }

  justUseAnotherNumber () {
    return this.anotherNumber * 3
  }

  @Resolver()
  throwSomeError () {
    throw new Error('simulated error')
  }
}

export class DemoServiceWithServiceInfo implements ServiceInfo {
  info (): { serviceName: string } {
    return {
      serviceName: 'CustomServiceName'
    }
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo2 (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}" - from echo2`
  }
}
