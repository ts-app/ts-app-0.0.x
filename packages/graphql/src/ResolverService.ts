import { Middleware, MiddlewareFunction } from '@ts-app/common'
import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'
import { ResolverError } from './ResolverError'
import { ResolverAuth } from './ResolverAuth'

let logError: (message?: any, ...optionalParams: any[]) => void = console.error
let _INSTANCE: ResolverService

/**
 * ResolverService is the core class that enables functions annotated with @Resolver() to work as
 * GraphQL resolvers.
 *
 * It is a singleton class where:
 * - Resolver() decorated functions are registered.
 * - All service classes with @Resolver() decorated functions must be are registered.
 * - Before/Afterwares can be registered to manage GraphQL queries from a centralized function.
 * - Graphql resolver functions are generated based on ResolverService's state.
 *
 */
export class ResolverService {
  private _serviceRegistry: {
    [className: string]: {
      service?: object
      resolvers: {
        [resolverName: string]: {
          descriptor: TypedPropertyDescriptor<Function>
          queryType: 'mutation' | 'query'
          paramNames: string[] | null
          auth: string[] | ResolverAuth
          name: string | null
        }
      }
    }
  }
  private beforewares: MiddlewareFunction<ResolverMiddlewareInput>[] = []
  private afterwares: MiddlewareFunction<ResolverMiddlewareInput>[] = []

  private constructor () {
    this._serviceRegistry = {}
  }

  static getInstance () {
    if (!_INSTANCE) {
      _INSTANCE = new ResolverService()
    }
    return _INSTANCE
  }

  /**
   * Register instance of service to handle resolver functions. There should only be one instance
   * of service class. Otherwise, it does not compute.
   *
   * @param {T} service service being registered
   * @param {string} className service name. Defaults to class name. WARNING: do not use this
   * parameter. It is here due to bad design/assumption. It should be removed.
   * @return {T} service being registered
   */
  registerService<T> (service: T, className = service.constructor.name): T {
    // --- support ServiceInfo
    if (typeof service.constructor.prototype.info === 'function') {
      const info = service.constructor.prototype.info
      className = info()[ 'serviceName' ]
    }

    if (className in this._serviceRegistry) {
      this._serviceRegistry[ className ].service = service
    } else {
      this._serviceRegistry[ className ] = { service, resolvers: {} }
    }

    return service
  }

  registerResolver (className: string, resolverName: string,
                    descriptor: TypedPropertyDescriptor<Function>, queryType: 'mutation' | 'query',
                    paramNames: string[] | null, auth: string[] | ResolverAuth, name: string | null) {
    let clazz = this._serviceRegistry[ className ]
    if (!clazz) {
      clazz = { resolvers: {} }
      this._serviceRegistry[ className ] = clazz
    }
    clazz.resolvers[ resolverName ] = { descriptor, queryType, paramNames, auth, name }
  }

  resetServices () {
    // leave "resolvers" alone. they are only set once (via @Resolver)

    // reset "service"
    Object.keys(this._serviceRegistry).map(className => {
      delete this._serviceRegistry[ className ].service
    })

    // reset before/after wares
    this.beforewares.length = 0
    this.afterwares.length = 0
  }

  /**
   * Returns a function that is binded to the service instance so that "this" works as expected.
   *
   * @param {string} className
   * @param {string} functionName
   * @return {any}
   */
  composeServiceFunction (className: string, functionName: string) {
    const descriptor = this._serviceRegistry[ className ].resolvers[ functionName ].descriptor
    const service = this._serviceRegistry[ className ].service
    const fn = descriptor.get ? descriptor.get() : descriptor.value
    return fn!.bind(service)
  }

  composeResolver (className: string, functionName: string, type: 'mutation' | 'query' = 'query', auth: string[] | ResolverAuth = [], resolverName: string = '') {
    // when resolver executes, it resolves the middleware
    return async (obj: any, args: any, context: any, info: any) => {
      try {
        const user = context.user || null
        const resolverParams = {
          obj, args, context, info,
          metadata: { className, functionName, resolverName, auth, type },
          user
        }
        const middleware = new Middleware<ResolverMiddlewareInput>([
          ...this.beforewares,
          this.createServiceFnMiddleware(className, functionName, resolverParams),
          ...this.afterwares
        ])

        return await middleware.resolve(resolverParams)
      } catch (e) {
        const now = Date.now()
        logError(`[${now}]`, e)
        if (e instanceof ResolverError) {
          return {
            error: e.message
          }
        } else {
          return {
            error: `Error processing resolver function [${className}.${functionName}] [${now}]`
          }
        }
      }
    }
  }

  makeResolvers () {
    const queryResolvers: { [resolverName: string]: any } = {}
    const mutationResolvers: { [resolverName: string]: any } = {}

    Object.keys(this._serviceRegistry).map(className => {
      const { resolvers, service } = this.serviceRegistry[ className ]
      if (service) {
        Object.keys(resolvers).map(key => {
          const { queryType, auth, name } = resolvers[ key ]
          // use custom resolver name if specified, else use decorated function's name
          const resolverName = name ? name : key
          if (queryType === 'mutation') {
            if (resolverName in mutationResolvers) {
              throw new Error(`Mutation resolver [${resolverName}] must be unique`)
            }
            mutationResolvers[ resolverName ] = this.composeResolver(className, key, 'query', auth, resolverName)
          } else {
            if (resolverName in queryResolvers) {
              throw new Error(`Query resolver [${resolverName}] must be unique`)
            }
            queryResolvers[ resolverName ] = this.composeResolver(className, key, 'mutation', auth, resolverName)
          }
        })
      }
    })

    return {
      Query: queryResolvers,
      Mutation: mutationResolvers
    }
  }

  setLogger ({ error }: { error: (message?: any, ...optionalParams: any[]) => void }) {
    logError = error
  }

  getService<T extends Object> (serviceType: Function | string): T | null {
    let serviceName

    if (typeof serviceType === 'string') {
      serviceName = serviceType
    } else if (typeof serviceType.prototype.info === 'function') {
      // --- support ServiceInfo
      serviceName = serviceType.prototype.info()[ 'serviceName' ]
    } else {
      serviceName = serviceType.name
    }

    // const serviceName = typeof serviceType === 'string' ? serviceType : serviceType.name
    const serviceObj = this._serviceRegistry[ serviceName ]
    if (serviceObj && serviceObj.service) {
      return serviceObj.service as T
    } else {
      return null
    }
  }

  registerBeforeware (before: MiddlewareFunction<ResolverMiddlewareInput>) {
    this.beforewares.push(before)
  }

  registerAfterware (after: MiddlewareFunction<ResolverMiddlewareInput>) {
    this.afterwares.push(after)
  }

  private createServiceFnMiddleware (className: string, functionName: string, resolverParams: object): MiddlewareFunction<ResolverMiddlewareInput> {
    // service function that performs business logic
    const serviceFn = this.composeServiceFunction(className, functionName)

    const resolver = this._serviceRegistry[ className ].resolvers[ functionName ]
    const { paramNames } = resolver
    if (paramNames) {
      // middleware transform args from object to array based on paramNames
      return ({ obj, args, context, info }, next) => {
        // convert args (object) to as array, pass to serviceFn as spread parameters
        const params = paramNames.map(name => args[ name ])
        next(serviceFn(...params, resolverParams))
      }
    } else {
      // middleware just pass args to serviceFn (it is expected to take a single object parameter)
      return ({ obj, args, context, info }, next) => {
        next(serviceFn(args, resolverParams))
      }
    }
  }

  get serviceRegistry () {
    return this._serviceRegistry
  }
}
