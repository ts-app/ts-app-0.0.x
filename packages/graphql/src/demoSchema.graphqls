type Query {
  echo (name: String!, message: String!): String
  throwSomeError: String
}

type Mutation {
  add (howMany: Int): Int
  promiseToSubtract (howMany: Int): Int
}
