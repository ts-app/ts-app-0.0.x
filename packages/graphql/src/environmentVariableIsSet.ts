import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'

/**
 * This function is used to enable/disable access to GraphQL resolver based on specified environment
 * variable to be set to '1'.
 *
 * @param {string} name Environment variable name
 * @return {(input: ResolverMiddlewareInput) => boolean}
 */
export const environmentVariableIsSet = (name: string) => {
  const authorized = process.env[ name ] === '1'

  return (input: ResolverMiddlewareInput): boolean => authorized
// tslint:disable-next-line
}
