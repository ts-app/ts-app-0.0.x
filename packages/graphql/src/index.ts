export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
