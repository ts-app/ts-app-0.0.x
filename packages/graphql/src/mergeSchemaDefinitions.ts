import { SchemaDefinition } from './SchemaDefinition'
import { mergeTypeDefs } from './mergeTypeDefs'

/**
 *
 * Merge an array of SchemaDefinition and return IExecutableSchemaDefinition (from Apollo) that can
 * be used as an input for makeExecutableSchema().
 *
 * It extracts contents from schema definition's "dependencies" to be returned as part of the
 * IExecutableSchemaDefinition. This allows reusing SchemaDefinition when specified as
 * "dependencies".
 *
 * @param {SchemaDefinition[]} definitions
 * @return {{typeDefs: string; resolvers: {Query: {}; Mutation: {}}}}
 */
export const mergeSchemaDefinitions = (definitions: SchemaDefinition[]) => {
  let typeDefs = ''
  let mergedResolvers = {
    Query: {},
    Mutation: {}
  }

  // flatten dependencies for all definitions
  const flatDependencies: SchemaDefinition[] = definitions.map(d => {
    if (d.dependencies) {
      return Object.keys(d.dependencies).map(name => d.dependencies![ name ])
    } else {
      return []
    }
  }).reduce((prev, current) => prev.concat(current), [])

  // resolve all definitions as "dependencies" plus provided "definitions"
  const allDefinitions = [ ...flatDependencies, ...definitions ]

  // perform merge
  allDefinitions.map(d => {
    // --- (simple) merge for typeDefs
    if (d.typeDefs) {
      if (typeDefs.length === 0) {
        typeDefs = d.typeDefs
      } else {
        typeDefs = mergeTypeDefs([ typeDefs, d.typeDefs ])
      }
    }

    if (d.resolvers) {
      mergedResolvers = {
        ...mergedResolvers,
        ...d.resolvers
      }

      if (d.resolvers.Query) {
        mergedResolvers = {
          ...mergedResolvers,
          Query: {
            ...mergedResolvers.Query,
            ...d.resolvers.Query
          }
        }
      }

      if (d.resolvers.Mutation) {
        mergedResolvers = {
          ...mergedResolvers,
          Mutation: {
            ...mergedResolvers.Mutation,
            ...d.resolvers.Mutation
          }
        }
      }
    }
  })

  return {
    typeDefs,
    resolvers: mergedResolvers
  }
  // tslint:disable-next-line
}
