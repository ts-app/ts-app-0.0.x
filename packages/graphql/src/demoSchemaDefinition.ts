import { loadFile } from '@ts-app/common'
import { DemoService } from './DemoService'
import { ResolverService } from './ResolverService'
import { standardSchemaDefinition } from './standardSchemaDefinition'
import { SchemaDefinition } from './SchemaDefinition'

export const demoSchemaDefinition = (): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  resolver.registerService(new DemoService())
  const typeDefs = loadFile(`${__dirname}/demoSchema.graphqls`)
  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
