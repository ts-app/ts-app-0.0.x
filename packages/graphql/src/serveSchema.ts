import * as bodyParser from 'body-parser'
import * as express from 'express'
import * as http from 'http'
import { Server } from 'http'
import { graphqlExpress, graphiqlExpress } from 'graphql-server-express'
import { GraphQLSchema } from 'graphql'
import chalk from 'chalk'
import { express as voyager } from 'graphql-voyager/middleware'
import { ResolverService } from './ResolverService'
import { SecurityMiddleware } from './SecurityMiddleware'

export const serveSchema = async ({ schema, uri = '/graphql', port = 3000, rootType, showInfo = true }: {
  schema: GraphQLSchema, uri?: string, port?: number, rootType?: string, showInfo?: boolean
}): Promise<Server> => {
  const app: any = express()

  const resolver = ResolverService.getInstance()
  const security = resolver.getService<SecurityMiddleware>('SecurityService')
  app.use(uri, bodyParser.json())
  if (security) {
    app.use(uri, security.middleware(), graphqlExpress(request => {
      const user = (request as any).user
      if (user) {
        return {
          schema,
          // --- request auth: step 3 of 3. assign user to graphql context
          context: { user }
        }
      } else {
        return { schema }
      }
    }))
  } else {
    app.use(uri, graphqlExpress({ schema }))
  }

  // TODO: secure graphiql
  app.use(`${uri}-ui`, graphiqlExpress({ endpointURL: uri }))

  // TODO: secure voyager
  app.use(`${uri}-diagram/Query`, voyager({
    endpointUrl: uri,
    displayOptions: { rootType: 'Query' }
  }))
  app.use(`${uri}-diagram/Mutation`, voyager({
    endpointUrl: uri,
    displayOptions: { rootType: 'Mutation' }
  }))
  if (rootType) {
    // diagram for an additional root type
    app.use(`${uri}-diagram`, voyager({
      endpointUrl: uri,
      displayOptions: { rootType }
    }))
  }

  const server = http.createServer(app)
  return new Promise<Server>(resolve => {
    server.listen(port, () => {
      const extraDiagram = rootType
        ? `\nDiagrams : ${chalk.bold('http://localhost:' + port + uri + '-diagram')} (Root Type: ${chalk.dim(rootType)})`
        : ''

      if (showInfo) {
        console.log(`GraphQL Server
==============
Endpoint : ${chalk.bold('http://localhost:' + port + uri)} ${!!security === false ? chalk.red('(without SecurityService)') : ''}
UI       : ${chalk.bold('http://localhost:' + port + uri + '-ui')}
Diagrams : ${chalk.bold('http://localhost:' + port + uri + '-diagram/Query')}
Diagrams : ${chalk.bold('http://localhost:' + port + uri + '-diagram/Mutation')}${extraDiagram}`)
      }
      resolve(server)
    })
  })
}
