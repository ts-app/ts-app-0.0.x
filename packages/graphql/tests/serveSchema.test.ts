import { ApolloFetch, createApolloFetch } from 'apollo-fetch'
import { Server } from 'http'
import gql from 'graphql-tag'
import { serveSchema } from '../src/serveSchema'
import { ResolverService } from '../src/ResolverService'
import { makeExecutableSchema } from 'graphql-tools'
import { demoSchemaDefinition } from '../src/demoSchemaDefinition'
import { mergeSchemaDefinitions } from '../src/mergeSchemaDefinitions'

describe('serveSchema', async () => {
  const resolverService = ResolverService.getInstance()
  let server: Server
  let fetch: ApolloFetch

  beforeAll(async () => {
    const demo = demoSchemaDefinition()
    const merged = mergeSchemaDefinitions([ demo ])
    const schema = makeExecutableSchema(merged)
    server = await serveSchema({ schema })
    fetch = createApolloFetch({ uri: 'http://localhost:3000/graphql' })
  })

  afterAll(async () => {
    await new Promise(resolve => server.close(resolve))

    resolverService.resetServices()
  })

  test('query echo()', async () => {
    const echo = await fetch({
      query: gql`query {
          echo(name: "Bob", message: "is jumping!")
      }`
    })
    expect(echo.errors).toBeUndefined()
    expect(echo.data).toMatchSnapshot()
  })

  test('mutate add() ', async () => {
    const add = await fetch({
      query: gql`mutation {
          add(howMany:12)
      }`
    })
    expect(add.errors).toBeUndefined()
    expect(add.data).toMatchSnapshot()
  })

  test('resolver info passed as last parameter', async () => {
    const resolverInfo = await fetch({
      query: gql`query {
          resolverInfo(title: "Candy") {
              title
              metadata user obj info args
          }
      }`
    })
    expect(resolverInfo.errors).toBeUndefined()
    expect(resolverInfo.data).toMatchSnapshot()
  })
})
