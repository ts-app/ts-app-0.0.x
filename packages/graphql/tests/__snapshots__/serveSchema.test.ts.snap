// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serveSchema mutate add()  1`] = `
Object {
  "add": 12,
}
`;

exports[`serveSchema query echo() 1`] = `
Object {
  "echo": "Hello, Bob! Message is \\"is jumping!\\"",
}
`;
